### PR TITLE
Let HLS Controller decide if subtitle should be burn in

### PR DESCRIPTION
**Pull request by gnattu**
Previously, we predicted whether the subtitle should be burned in with transcode reasons, but that was not accurate because the actual transcoding codec is only determined after the client has requested the stream. This pass through the option to the `DynamicHlsController` to handle the subtitle burn-in during the actual transcoding process. Now the client should be responsible to conditionally load the subtitle when this option is enabled.

**Changes**

**Issues**

### CODE REVIEW

File: Jellyfin.Api/Controllers/DynamicHlsController.cs [hunk 1]
```diff
@@ -251,7 +253,7 @@ [FromQuery] [RegularExpression(EncodingHelper.ValidationRegex)] string? subtitle
             Height = height,
             VideoBitRate = videoBitRate,
             SubtitleStreamIndex = subtitleStreamIndex,
-            SubtitleMethod = subtitleMethod ?? SubtitleDeliveryMethod.Encode,
+            SubtitleMethod = subtitleMethod ?? SubtitleDeliveryMethod.External,
```
Comment (dmitrylyzo) [inline Jellyfin.Api/Controllers/DynamicHlsController.cs:256]: I think the idea is that `Encode` will always work.
But `SubtitleMethod` must be specified for stable results, though.

**Review by Shadowghost — APPROVED** [review]

File: Jellyfin.Api/Controllers/DynamicHlsController.cs [hunk 2]
```diff
@@ -629,7 +636,8 @@ [FromQuery] [RegularExpression(EncodingHelper.ValidationRegex)] string? subtitle
         [FromQuery] EncodingContext? context,
         [FromQuery] Dictionary<string, string> streamOptions,
         [FromQuery] bool enableAdaptiveBitrateStreaming = true,
-        [FromQuery] bool enableAudioVbrEncoding = true)
+        [FromQuery] bool enableAudioVbrEncoding = true,
+        [FromQuery] bool alwaysBurnInSubtitleWhenTranscoding = false)
```
Comment (dmitrylyzo) [inline Jellyfin.Api/Controllers/DynamicHlsController.cs:640]: Do we need `alwaysBurnInSubtitleWhenTranscoding` in the audio enpoints?

Comment (gnattu) [reply Jellyfin.Api/Controllers/DynamicHlsController.cs:640]: Probably not, but there are also other video options in audio endpoints

**Review by crobibero — APPROVED** [review]

File: Jellyfin.Api/Controllers/DynamicHlsController.cs (hunk 1 above)
Comment (gnattu) [reply Jellyfin.Api/Controllers/DynamicHlsController.cs:256]: The problem is that StreamInfo's URL builder always specifies Encode and will leave it blank for External. We should make the behavior consistent.

File: Jellyfin.Api/Controllers/DynamicHlsController.cs [hunk 3]
```diff
@@ -911,6 +921,7 @@ [FromQuery] [RegularExpression(EncodingHelper.ValidationRegex)] string? subtitle
     /// <param name="context">Optional. The <see cref="EncodingContext"/>.</param>
     /// <param name="streamOptions">Optional. The streaming options.</param>
     /// <param name="enableAudioVbrEncoding">Optional. Whether to enable Audio Encoding.</param>
+    /// <param name="alwaysBurnInSubtitleWhenTranscoding">Whether to always burn in subtitles when transcoding.</param>
```
Comment (dmitrylyzo) [inline Jellyfin.Api/Controllers/DynamicHlsController.cs:924]: Then remove this one as well.
_I expected you to remove them all._ :sweat_smile: 
```suggestion
```

File: Jellyfin.Api/Controllers/DynamicHlsController.cs [hunk 4]
```diff
@@ -966,7 +977,8 @@ [FromQuery] [RegularExpression(EncodingHelper.ValidationRegex)] string? subtitle
         [FromQuery] int? videoStreamIndex,
         [FromQuery] EncodingContext? context,
         [FromQuery] Dictionary<string, string> streamOptions,
-        [FromQuery] bool enableAudioVbrEncoding = true)
+        [FromQuery] bool enableAudioVbrEncoding = true,
+        [FromQuery] bool alwaysBurnInSubtitleWhenTranscoding = false)
```
Comment (dmitrylyzo) [inline Jellyfin.Api/Controllers/DynamicHlsController.cs:981]:
```suggestion
        [FromQuery] bool enableAudioVbrEncoding = true)
```

File: Jellyfin.Api/Controllers/DynamicHlsController.cs [hunk 5]
```diff
@@ -1018,7 +1030,8 @@ [FromQuery] [RegularExpression(EncodingHelper.ValidationRegex)] string? subtitle
             VideoStreamIndex = videoStreamIndex,
             Context = context ?? EncodingContext.Streaming,
             StreamOptions = streamOptions,
-            EnableAudioVbrEncoding = enableAudioVbrEncoding
+            EnableAudioVbrEncoding = enableAudioVbrEncoding,
+            AlwaysBurnInSubtitleWhenTranscoding = alwaysBurnInSubtitleWhenTranscoding
```
Comment (dmitrylyzo) [inline Jellyfin.Api/Controllers/DynamicHlsController.cs:1034]:
```suggestion
            AlwaysBurnInSubtitleWhenTranscoding = false
```

File: Jellyfin.Api/Controllers/DynamicHlsController.cs [hunk 6]
```diff
@@ -1267,6 +1283,7 @@ [FromQuery] [RegularExpression(EncodingHelper.ValidationRegex)] string? subtitle
     /// <param name="context">Optional. The <see cref="EncodingContext"/>.</param>
     /// <param name="streamOptions">Optional. The streaming options.</param>
     /// <param name="enableAudioVbrEncoding">Optional. Whether to enable Audio Encoding.</param>
+    /// <param name="alwaysBurnInSubtitleWhenTranscoding">Whether to always burn in subtitles when transcoding.</param>
```
Comment (dmitrylyzo) [inline Jellyfin.Api/Controllers/DynamicHlsController.cs:1286]:
```suggestion
```

File: Jellyfin.Api/Controllers/DynamicHlsController.cs [hunk 7]
```diff
@@ -1328,7 +1345,8 @@ [FromQuery] [RegularExpression(EncodingHelper.ValidationRegex)] string? subtitle
         [FromQuery] int? videoStreamIndex,
         [FromQuery] EncodingContext? context,
         [FromQuery] Dictionary<string, string> streamOptions,
-        [FromQuery] bool enableAudioVbrEncoding = true)
+        [FromQuery] bool enableAudioVbrEncoding = true,
+        [FromQuery] bool alwaysBurnInSubtitleWhenTranscoding = false)
```
Comment (dmitrylyzo) [inline Jellyfin.Api/Controllers/DynamicHlsController.cs:1349]:
```suggestion
        [FromQuery] bool enableAudioVbrEncoding = true)
```

File: Jellyfin.Api/Controllers/DynamicHlsController.cs [hunk 8]
```diff
@@ -1382,7 +1400,8 @@ [FromQuery] [RegularExpression(EncodingHelper.ValidationRegex)] string? subtitle
             VideoStreamIndex = videoStreamIndex,
             Context = context ?? EncodingContext.Streaming,
             StreamOptions = streamOptions,
-            EnableAudioVbrEncoding = enableAudioVbrEncoding
+            EnableAudioVbrEncoding = enableAudioVbrEncoding,
+            AlwaysBurnInSubtitleWhenTranscoding = alwaysBurnInSubtitleWhenTranscoding
```
Comment (dmitrylyzo) [inline Jellyfin.Api/Controllers/DynamicHlsController.cs:1404]:
```suggestion
            AlwaysBurnInSubtitleWhenTranscoding = false
```

**Review by dmitrylyzo — APPROVED** [review]

**Review by rehan678p — APPROVED** [review]

**Pull request merged**